File: coding-practice/LeetCode 75/Easy/13 - Roman To Integer/Python/Solution.py
roman_values = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}

class Solution:
    def __init__(self) -> None:
        pass

    def myRomanToInt(self, s):
        total = 0
        for i in range(len(s)):
            current_value = roman_values[s[i]]
            if i + 1 == len(s):
                total += current_value        
            elif s[i] == 'I' and (s[i + 1] == 'V' or s[i + 1] == 'X'):
                total -= 1
            elif s[i] == 'X' and (s[i + 1] == 'L' or s[i + 1] == 'C'):
                total -= 10
            elif s[i] == 'C' and (s[i + 1] == 'D' or s[i + 1] == 'M'):
                total -= 100
            else:
                total += current_value
        return total
    
class ChatGPTSolution:
    def __init__(self) -> None:
        pass

    def romanToInt(self, s):
        total = 0
        prev_value = 0

        for char in reversed(s):
            current_value = roman_values[char]

            if current_value < prev_value:
                total -= current_value
            else:
                total += current_value
            
            prev_value = current_value
        return total



s = 'MCMXCIV'
obj = ChatGPTSolution()
myObj = Solution()
romanInt = obj.romanToInt(s)
n = myObj.myRomanToInt(s)
print(romanInt, n)

File: coding-practice/LeetCode 75/Easy/13 - Roman To Integer/C++/Solution.cpp
#include<iostream>
#include<vector>
#include<map>

using namespace std;

class Solution{
	private:
		map<char, int> romanValues;	
	
	public:	
	Solution(){
		romanValues.insert(make_pair('I', 1));
	    romanValues.insert(make_pair('V', 5));
	    romanValues.insert(make_pair('X', 10));
	    romanValues.insert(make_pair('L', 50));
	    romanValues.insert(make_pair('C', 100));
	    romanValues.insert(make_pair('D', 500));
	    romanValues.insert(make_pair('M', 1000));
	}
    
    int romanToInt(string nums){
    	int total = 0;
    	for(int i = 0; i < nums.size(); i++){
    		int current_value = romanValues[nums[i]];
    		if(i + 1 == nums.size()){
    			total += current_value;
			}
			else if(nums[i] == 'I' and (nums[i + 1] == 'X' or nums[i + 1] == 'V')){
				total -= 1;
			}
			else if(nums[i] == 'X' and (nums[i + 1] == 'L' or nums[i + 1] == 'C')){
				total -= 10;
			}
			else if(nums[i] == 'C' and (nums[i + 1] == 'D' or nums[i + 1] == 'M')){
				total -= 100;
			}
			else{
				total += current_value;
			}
		}
		
		return total;
	};
};


int main(){
	Solution obj;
	string roman = "MCMXCIV";
	int romanInt = obj.romanToInt(roman);
	cout << "The integer value of " << roman << " is " << romanInt << endl;
    
	/*
	for(int i = 0; i < s.size(); i++){
		cout << s[i] << endl;
	}
	*/
	return 0;
};
